compiler/src/main/java/com/squareup/anvil/compiler/codegen/CodeGenerationExtension.kt

```diff
@@ -0,0 +1,97 @@
+package com.squareup.anvil.compiler.codegen
+
+import com.squareup.anvil.compiler.codegen.CodeGenerator.GeneratedFile
+import org.jetbrains.kotlin.analyzer.AnalysisResult
+import org.jetbrains.kotlin.analyzer.AnalysisResult.RetryWithAdditionalRoots
+import org.jetbrains.kotlin.com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.com.intellij.psi.PsiManager
+import org.jetbrains.kotlin.com.intellij.testFramework.LightVirtualFile
+import org.jetbrains.kotlin.container.ComponentProvider
+import org.jetbrains.kotlin.context.ProjectContext
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.idea.KotlinFileType
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.resolve.BindingTrace
+import org.jetbrains.kotlin.resolve.jvm.extensions.AnalysisHandlerExtension
+import java.io.File
+
+internal class CodeGenerationExtension(
+  private val codeGenDir: File
+) : AnalysisHandlerExtension {
+
+  private var didRecompile = false
+
+  private val codeGenerators = mutableListOf<CodeGenerator>()
+
+  fun registerCodeGenerator(codeGenerator: CodeGenerator) {
```

Comment:
If you didn't already know to call this, it'd be easy to forget to call `codeGenerationExtension.registerCodeGenerator(ContributesToGenerator())`. Would it be more beneficial to enforce that via a constructor argument instead?

Reply:
Hm, I like this better. Constructors should be free of any logic.

Reply:
I'm asking because I don't see where unregisterCodeGenerator would be used, so getting rid of the internally managed list and instead pass in when you initialize the class. I might be missing something if those are used outside the scope of this PR though.

Reply:
That's fair, I like that actually better.

Reply:
There's no logic? You're injecting the list of codeGenerators instead of relying on someone to call that method beforehand. Without it this class won't work properly, right? Your call.

Reply:
Changed